test/test_log_forwarding.py

```diff
@@ -0,0 +1,159 @@
+import gzip
+from io import (
+    StringIO,
+)
+import json
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+import chalice.app
+from moto import (
+    mock_s3,
+)
+
+from app_test_case import (
+    LambdaEnvTestCase,
+)
+from azul import (
+    cached_property,
+    config,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.indexer.log_forwarding_controller import (
+    LogForwardingController,
+)
+from azul.service.storage_service import (
+    StorageService,
+)
+
+
+class TestLogForwarding(LambdaEnvTestCase):
+
+    @property
+    def log_bucket(self) -> str:
+        return aws.qualified_bucket_name(config.logs_term)
+
+    @property
+    def log_file_key(self) -> str:
+        prefix = config.alb_access_log_path_prefix(config.gitlab_term)
+        return '/'.join([prefix, 'AWSLogs', '123123123123', 'elasticloadbalancing', '2023', '01', '01', 'test.log.gz'])
+
+    @cached_property
+    def storage_service(self) -> StorageService:
+        return StorageService(bucket_name=self.log_bucket)
+
+    @mock_s3
+    def test(self):
+        self.storage_service.create_bucket()
+        event = chalice.app.S3Event(context={}, event_dict={
+            'Records': [{
+                's3': {
+                    'bucket': {'name': self.log_bucket},
+                    'object': {'key': self.log_file_key}
+                }
+            }]
+        })
+
+        log_escape_sequences_by_input = {
+            # Quotation marks are escaped because they are used wrap fields that
+            # may contain spaces
+            '"': r'\x22',
+            # Backslashes are escaped because they are used to escape non-ASCII
+            # characters (see subsequent cases)
+            '\\': r'\x5c',
+            # Non-ASCII character (Latin-1)
+            'È': r'\xC3\x88',
+            # Non-ASCII character (Cyrillic)
+            'Ё': r'\xD0\x81',
+            # Non-ASCII character (Emoticon)
+            '😀': r'\xf0\x9f\x98\x80',
+        }
+
+        for unescaped, escaped in log_escape_sequences_by_input.items():
+            with self.subTest(chars=escaped):
+                input = [' '.join([
+                    'https',
+                    '2022-12-31T23:55:00.388951Z',
+                    'app/azul-gitlab-alb/c051f98624e68d7e',
+                    '172.71.0.215:36056', '172.71.0.215:80',
+                    '0.000', '0.002', '0.000',
+                    '204', '204',
+                    '963', '229',
+                    '"POST '
+                    f'https://gitlab.dev.singlecell.gi.ucsc.edu:443/api/v4/jobs/request?chars={escaped} '
+                    'HTTP/1.1"',
```

Comment:
Did you check whether curl sends UTF-8 with HTTP/1.1?

Reply:
Because the non-ASCII characters used in the previous experiments were generated using python, they were UTF8-encoded before being received by cURL. For the new experiment, I took measures to write the bytes for a latin1-encoded `'¢'`symbol (`\xa2`` in latin1, `\xc2\xa2` in utf8). Here's the cURL invocation: 

```
$ curl -vv --http1.1 -A "encoding-experiment $(python3 -c 'import sys; sys.stdout.buffer.write(chr(0xA2).encode("lahttps://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/EXPERIMENT
*   Trying 172.71.1.37:443...
* Connected to gitlab.dev.singlecell.gi.ucsc.edu (172.71.1.37) port 443 (#0)
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: CN=gitlab.dev.singlecell.gi.ucsc.edu
*  start date: Feb 22 00:00:00 2023 GMT
*  expire date: Dec  7 23:59:59 2023 GMT
*  subjectAltName: host "gitlab.dev.singlecell.gi.ucsc.edu" matched cert's "gitlab.dev.singlecell.gi.ucsc.edu"
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M02
*  SSL certificate verify ok.
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> GET /ucsc/EXPERIMENT HTTP/1.1
> Host: gitlab.dev.singlecell.gi.ucsc.edu
> User-Agent: encoding-experiment �
> Accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Mark bundle as not supporting multiuse
< HTTP/1.1 302 Found
< Date: Thu, 23 Feb 2023 05:54:31 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 121
< Connection: keep-alive
< Server: nginx
< Cache-Control: no-cache
< Content-Security-Policy: 
< Location: https://gitlab.dev.singlecell.gi.ucsc.edu/users/sign_in
< Permissions-Policy: interest-cohort=()
< Pragma: no-cache
< Set-Cookie: _gitlab_session=10b2dfd93e44ad991df9e3e0aa6effde; path=/; expires=Thu, 23 Feb 2023 07:54:31 GMT; secure; HttpOnly; SameSite=None
< X-Content-Type-Options: nosniff
< X-Download-Options: noopen
< X-Frame-Options: SAMEORIGIN
< X-Permitted-Cross-Domain-Policies: none
< X-Request-Id: 01GSYDCSS335XGP65ZMH4SX2D6
< X-Runtime: 0.086839
< X-Ua-Compatible: IE=edge
< X-Xss-Protection: 1; mode=block
< Strict-Transport-Security: max-age=63072000
< Referrer-Policy: strict-origin-when-cross-origin
< 
* Connection #0 to host gitlab.dev.singlecell.gi.ucsc.edu left intact
* ```
Note that the `¢` is rendered as `�`, presumably because my console uses utf8. I think this is evidence that cURL does not apply UTF8 encoding by itself.

The result of this latest experiment is that the special character appeared in the S3 logs as the escape sequence `\xA2`, i.e., the exact sequence of bytes that present in the request, with no UTF8 encoding. This crashed the log forwarder, presumably because our current implementation assumes that the escaped sequence can be decoded to UTF8. Unfortunately, the original exception did not appear in the log message, so that will also need to be fixed.